### PR TITLE
Feature/pypika

### PR DESCRIPTION
**Pull request by akmiller01**
Starting a PR for this branch now @Napho . Since we've both worked on it we'll rely on @edwinmp to do a review.

This afternoon I fixed the sql_load to match our new models and made a new fixture out of it. Then I added a fixture load test case to make sure CI alerts us if we ever break fixtures in the future.

I've modeled the `query_kwargs` how I imagine they'll end up, but feel free to correct.

Also, first draft PR I've ever made, so let's see how this works out!

### CODE REVIEW

**Comment by akmiller01** [comment]
Holy cow it's working @Napho & @edwinmp :

```
python manage.py shell
>>> from core.models import Operation
>>> o = Operation.objects.first()
>>> o.build_query()
'SELECT * FROM "repo.crs_current" WHERE "donor_code"=2000 AND "ftr"=\'Badman\''
```

**Comment by akmiller01** [comment]
Added an aggregate step second and fixed a bug. Now it can do aggregate and filter subsequently:

```
>>> o.build_query()
'SELECT "sq0"."id","sq0"."value",SUM("sq0"."value") "value_Sum" FROM (SELECT * FROM "repo.crs_current" WHERE "donor_code"=2000 AND "ftr"=\'Badman\') "sq0" GROUP BY "sq0"."id","sq0"."value"'
```

I'll leave the rest to you @Napho :P I was going to start on loading the CRS test data as a fixture, but I couldn't help but test the test metadata once I fixed it.

One thing I think this is missing now is in the `__init__` we should make sure to order the `query_steps` by `step_id`

**Review by Napho — CHANGES_REQUESTED** [review]
@edwinmp  and @akmiller01 if you look at the window function, there is an explicit check on some of the function being called, i think its done a bit poorly but im not sure what the best way to handle it would be. I feel like it gives a bit of dependency on @edwinmp to know what is happening at pypika; which shouldn't exactly be the case. Your thoughts?

**Comment by edwinmp** [comment]
Haven't looked critically at the code, but I agree with your last statement. It should be abstracted enough that `pypika` shouldn't dictate how I get data to it & the shape it'll take. It should be designed in such a way that we can want to replace `pypika` next year and nothing would changed on the frontend. @akmiller01 @Napho

**Comment by akmiller01** [comment]
Hey @edwinmp , just pushed a feature to this branch for data serialization. If you have already 

```
python manage.py loaddata test_data
python manage.py loaddata --database=datasets test_datasets
```

Then you should be able to get a Token via the instructions in README.md and make a Postman query to `localhost:8000/api/operation_data/1` with 1 being the PK of the test_data operation. It turns a JSON object with `count (int)`, `columns (array)`, and `data (array of arrays)`.
E.g.
image


This endpoint defaults to 10 rows. You can also get a full output by adding the parameter `full=true` to get the full unlimited output.
E.g.
image

Do you think this is sufficient? Or should we just pass something like `limit` as a param instead and let `limit=false` stand for full output?

**Comment by akmiller01** [comment]
Oh also, on replacing Pypika. I think if we really want to manage our dependency on it, we just need to enumerate and map the functions we're using. For e.g. this line, instead of looking up a full list of functions from Pypika https://github.com/devinit/ddw-analyst-ui/blob/feature/pypika/core/pypika_utils.py#L34

We could just make a dictionary mapping a finite set of functions to their equivalents in Pypika. It wouldn't save us headache if a future package needs arguments fed to sql functions in a completely different way, but it would allow is to explicitly map functions without relying on one exactly being called "Sum" for e.g.

**Comment by edwinmp** [comment]
> Do you think this is sufficient? Or should we just pass something like `limit` as a param instead and let `limit=false` stand for full output?

Think it needs legit pagination i.e a `limit`, `offset/page` (I prefer page) and a `count`. Count should hold the total number of records, even if you only returned 10. Excluding the `limit` should return the full data.

Also not quite following this `operation_data` example ... seems to be fetching the data of an already created operation. Guess it'll make better sense once I get the hang of how an operation is created. We shall discuss that in the meeting tomorrow

**Comment by akmiller01** [comment]
Working on that count feature now @edwinmp .

This test shows how you can make an Operation with Steps in an API POST: https://github.com/devinit/ddw-analyst-ui/blob/feature/pypika/core/tests.py#L69-L74

And this test shows how you how the models are working:
https://github.com/devinit/ddw-analyst-ui/blob/feature/pypika/core/tests.py#L85-L97

**Comment by akmiller01** [comment]
Ask and ye shall receive @edwinmp 

image

Params are now `limit` (default 10) `page` (default 1) and `full` (overrides limit and page for a complete dataset).

**Comment by edwinmp** [comment]
@akmiller01 rebased `develop` in here ... a few conflicts that I dealt with.

**Comment by Napho** [comment]
Additional comments follwing https://github.com/devinit/ddw-analyst-ui/pull/60/commits/48c549e77218854fd884e11dc57d57ff3bcb0b23

Hey @Naphlin, I had a stab at finishing off our functions with `transform.` It still needs testing, and I had to split it into `scalar_transform` and `multi_transform` depending on whether it's operating on several columns, or one column and one scalar value. https://github.com/devinit/ddw-analyst-ui/commit/e892ae48b73a6ef3dca9fed1c9949ba35637a358
But that being said, I think we've got a whole can'o'worms we haven't even opened yet with deflation/conversion. Where we've been able to be abstract so far, even with things like text searches in `transform`, deflation/conversion is going to need to be very specific, and will involve look-ups from tables it's expecting to exist

----

Alex [9:41 PM]
Also refactored `filter` so that users can do more than just equalities. I've got the mapping in the `pypika_utils` function, but we can now do equal, less than, greater than, etc. Plus a text search with ilike
So check out how the expected output from this test looks now

----
The only remaining problem with `filter` is intuiting whether a User wants AND filters or OR filters. I'm tempted to go with OR as the default, since if they want AND, they can just run multiple filters instead of one. And right now this structure wouldn't allow the CRS bundle to be built in one fell-swoop `("short_description" ILIKE ...) OR ("long_description" ILIKE ...)` Thoughts?

**Comment by akmiller01** [comment]
Does the OR combination I proposed work for you @Napho ?

**Comment by Napho** [comment]
> Does the OR combination I proposed work for you @Napho ?

@akmiller01  it does work, ive made a few changes in the course of my test though, not more than 10 or so lines, will push past 5

**Comment by akmiller01** [comment]
@edwinmp Python coverage seems to be working now. I'm not sure our biggest coverage offenders like `views.py` are good candidates for unit testing, though...

**Comment by akmiller01** [comment]
@Napho finishing off some tests, it seems like `MEDIAN` isn't a built in function in psql:
```
ERROR:  function median(unknown) does not exist
LINE 1: SELECT "sq0".*,MEDIAN('usd_commitment') OVER(PARTITION BY ye...
                       ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

**Comment by akmiller01** [comment]
Actually, `AVG` and `STDDEV` fail in PSQL for unknown reasons as well...
```
analyst_ui=# SELECT "sq0".*,AVG('usd_disbursement_defl') OVER(PARTITION BY year) FROM (SELECT * FROM "repo"."crs_current") "sq0";
ERROR:  function avg(unknown) is not unique
LINE 1: SELECT "sq0".*,AVG('usd_disbursement_defl') OVER(PARTITION B...
                       ^
HINT:  Could not choose a best candidate function. You might need to add explicit type casts.
analyst_ui=# SELECT "sq0".*,STDDEV('usd_disbursement_defl') OVER(PARTITION BY year) FROM (SELECT * FROM "repo"."crs_current") "sq0";
ERROR:  invalid input syntax for type double precision: "usd_disbursement_defl"
LINE 1: SELECT "sq0".*,STDDEV('usd_disbursement_defl') OVER(PARTITIO...
```

I'm wondering if these Window function's couldn't also be accomplished by the other functions in `aggregate`? It looks like pypika.functions has most bases covered: https://pypika.readthedocs.io/en/latest/api/pypika.functions.html

**Comment by akmiller01** [comment]
Hey @edwinmp , with @Napho off today, do you think it makes sense to shelve the window function for now and get a review of this branch?

**Comment by edwinmp** [comment]
@akmiller01 not sure what the `window` function is meant for, but if there's enough complete & stable within this branch for a review & merge, then sure... don't see why not.

**Comment by akmiller01** [comment]
@edwinmp , I don't think we can merge without Naphlin today anyway. He's got changes requested and we cannot dismiss them without disabling branch protection first. But I think the window function implements this behavior https://www.postgresql.org/docs/9.1/tutorial-window.html

It's just that the functions I think @Napho was hoping to use with it like `AVG` `MEDIAN` and `STDDEV` don't appear to exist natively in psql. Once I've sorted out test coverage, I'm going to see if there's a way to sort that out...

**Comment by edwinmp** [comment]
Definitely don't merge yet... would like to do a review as well. :)

**Comment by akmiller01** [comment]
Breakthrough:
```
analyst_ui=# select AVG('usd_commitment') from "repo"."crs_current";
ERROR:  function avg(unknown) is not unique
LINE 1: select AVG('usd_commitment') from "repo"."crs_current";
               ^
HINT:  Could not choose a best candidate function. You might need to add explicit type casts.
analyst_ui=# select AVG(usd_commitment) from "repo"."crs_current";
        avg         
--------------------
 9.2030000000000000
(1 row)
```
The window function must be passing column names as values rather than fields...

**Comment by akmiller01** [comment]
Ready for review @edwinmp and @Napho

**Comment by akmiller01** [comment]
Hey @napho for when we get more than just OECD data in here, will we ever need to join on differing column names? Like `donor_code=fts_donor`? If so we may want to make the join_on object a dict instead of list.

**Review by edwinmp — CHANGES_REQUESTED** [review]
Made a few changes to minor things I could follow... tested, but still, have a look to be sure I've not broken anything. If I have, don't hesitate to revert.

File: core/const.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# Default value for limit if its not specified in the query
+default_limit_count = 10
```
[the remark this answers is not in the document]
Comment (Napho) [reply core/const.py]: From what i see in settings file, that CAPS could be the right convention here too, changed accordingly

**Review by Napho — DISMISSED** [review]

**Review by edwinmp — DISMISSED** [review]

**Review by Napho — APPROVED** [review]

**Pull request merged**